### PR TITLE
Open telemetry

### PR DESCRIPTION
**Pull request by genevieveluyt**
Currently supports stdout as a provider for both tracing and metrics and Prometheus for metrics.

## Tracing

All HTTP requests to the key retrieval/submission service produce a trace. eg.

```
key-retrieval  | 	"SpanContext": {
key-retrieval  | 		"TraceID": "eb7083a99a159bc3624df6bf8a859589",
key-retrieval  | 		"SpanID": "2df0fb1627703baf",
key-retrieval  | 		"TraceFlags": 1
key-retrieval  | 	},
key-retrieval  | 	"ParentSpanID": "0000000000000000",
key-retrieval  | 	"SpanKind": 1,
key-retrieval  | 	"Name": "HTTP Request",
key-retrieval  | 	"StartTime": "2020-05-23T15:02:24.780002-04:00",
key-retrieval  | 	"EndTime": "2020-05-23T15:02:24.780022942-04:00",
key-retrieval  | 	"Attributes": [
key-retrieval  | 		{
key-retrieval  | 			"Key": "http.url",
key-retrieval  | 			"Value": {
key-retrieval  | 				"Type": "STRING",
key-retrieval  | 				"Value": "/services/ping"
key-retrieval  | 			}
key-retrieval  | 		}
key-retrieval  | 	],
key-retrieval  | 	"MessageEvents": null,
key-retrieval  | 	"Links": null,
key-retrieval  | 	"StatusCode": 0,
key-retrieval  | 	"StatusMessage": "",
key-retrieval  | 	"HasRemoteParent": false,
key-retrieval  | 	"DroppedAttributeCount": 0,
key-retrieval  | 	"DroppedMessageEventCount": 0,
key-retrieval  | 	"DroppedLinkCount": 0,
key-retrieval  | 	"ChildSpanCount": 0,
key-retrieval  | 	"Resource": null
```

We probably want to exclude the `/ping` endpoint but for now it makes testing easy.

## Metrics

Publishes system stats periodically. The metrics are always an aggregate (eg. min / max / sum) but since the count is set to 1 this always represents a single value (ie. min == max == sum == actual value)

Current metrics:
* total memory
* used memory
* used memory as a percentage
* available memory
* CPU usage as a percentage

See [gopsutil](https://godoc.org/github.com/shirou/gopsutil) for more metrics we could add

```
key-retrieval  | {
key-retrieval  | 	"time": "2020-05-25T18:55:37.209771-04:00",
key-retrieval  | 	"updates": [
key-retrieval  | 		{
key-retrieval  | 			"name": "covidshield.system.memory.total",
key-retrieval  | 			"min": 34359738368,
key-retrieval  | 			"max": 34359738368,
key-retrieval  | 			"sum": 34359738368,
key-retrieval  | 			"count": 1
key-retrieval  | 		},
key-retrieval  | 		{
key-retrieval  | 			"name": "covidshield.system.memory.usedpercent",
key-retrieval  | 			"min": 55.67512512207031,
key-retrieval  | 			"max": 55.67512512207031,
key-retrieval  | 			"sum": 55.67512512207031,
key-retrieval  | 			"count": 1
key-retrieval  | 		},
key-retrieval  | 		{
key-retrieval  | 			"name": "covidshield.system.memory.used",
key-retrieval  | 			"min": 19129827328,
key-retrieval  | 			"max": 19129827328,
key-retrieval  | 			"sum": 19129827328,
key-retrieval  | 			"count": 1
key-retrieval  | 		},
key-retrieval  | 		{
key-retrieval  | 			"name": "covidshield.system.memory.free",
key-retrieval  | 			"min": 15229911040,
key-retrieval  | 			"max": 15229911040,
key-retrieval  | 			"sum": 15229911040,
key-retrieval  | 			"count": 1
key-retrieval  | 		},
key-retrieval  | 		{
key-retrieval  | 			"name": "covidshield.system.cpu.percent",
key-retrieval  | 			"min": 5.1257439047801885,
key-retrieval  | 			"max": 5.1257439047801885,
key-retrieval  | 			"sum": 5.1257439047801885,
key-retrieval  | 			"count": 1
key-retrieval  | 		}
key-retrieval  | 	]
key-retrieval  | }
```

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,24 +1,40 @@
-# Contributing
+# Contributing Guidelines
 
-Thank you for considering contributing to COVID Shield!
+Welcome to CovidShield. Thank you for considering contributing to COVID Shield!
```
Comment (Ginja) [inline CONTRIBUTING.md:3]: ~CovidShield~ COVID Shield.

However, it seems like `Welcome to CovidShield` is redundant with the next sentence?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -93,3 +116,45 @@ Finally, run:
 ```shell
 $ make test
 ```
+
+## Metrics and Tracing
+
+CovidShield uses [OpenTelemetry](https://github.com/open-telemetry/opentelemetry-go) to configure the metrics and tracing for the server, both the key retrieval and key submission.
```
Comment (Ginja) [inline CONTRIBUTING.md:122]:
```suggestion
COVID Shield uses [OpenTelemetry](https://github.com/open-telemetry/opentelemetry-go) to configure the metrics and tracing for the server, both the key retrieval and key submission.
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -69,7 +92,7 @@ PORT=8001 ./key-retrieval
 
 Note that 302 is a [MCC](https://www.mcc-mnc.com/): 302 represents Canada.
 
-### Running Tests
+## Testing
 
 If you're at Shopify, `dev up` will configure the database for you. If not
 you will need to point to your database server using the environment variables
```
Comment (Ginja) [inline CONTRIBUTING.md:98]: Did we want to remove all Shopify-specific tooling from this repo? There was talks of it before.

Cc: @burke

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,24 +1,40 @@
...
+Welcome to CovidShield. Thank you for considering contributing to COVID Shield!
 
 We’d love to get your issues (if you find any bugs) and PRs (if you have any fixes)!
 
+- [Code of Conduct](#code-of-conduct)
+- [Reporting Security Issues](#reporting-security-issues)
+- [Contributing](#contributing)
+  - [Contributing Documentation](#contributing-documentation)
+  - [Contributing Code](#contributing-code)
+- [Running CovidShield](#running-covidshield)
```
Comment (Ginja) [inline CONTRIBUTING.md:12]:
```suggestion
- [Running COVID Shield](#running-covidshield)
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -28,7 +44,9 @@ as you'd expect.
 Once you're happy with your changes, please fork the repository and push your
 code to your fork, then open a PR against this repository.
 
-## Development Environment via docker-compose
+## Running CovidShield
```
Comment (Ginja) [inline CONTRIBUTING.md:47]:
```suggestion
## Running COVID Shield
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -93,3 +116,45 @@ Finally, run:
...
+
+If you want to use Prometheus, please see the additional configuration requirements below.
+
+#### Prometheus 
+
+In order to use Prometheus as a metrics solution, you'll need to be running it in your environment. 
+
+You can follow the instructions [here](https://prometheus.io/docs/prometheus/latest/installation/) for running Prometheus. 
+
+You will need to edit the configuration file, `prometheus.yml` to add an additional target so it actually polls the metrics coming from the CovidShield server:
```
Comment (Ginja) [inline CONTRIBUTING.md:140]:
```suggestion
You will need to edit the configuration file, `prometheus.yml` to add an additional target so it actually polls the metrics coming from the COVID Shield server:
```

File: CONTRIBUTING.md (hunk 3 above)
Comment (burke) [reply CONTRIBUTING.md:98]: Yeah it probably does make sense to st this point.

File: pkg/telemetry/telemetry.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	if tracerProvider == "" {
+		log(nil, nil).Info("TRACER_PROVIDER not set, tracing will not be generated.")
+		return cleanupFunc
+	}
+
+	var exporter *tracerstdout.Exporter
+	var err error
+	switch tracerProvider {
+	case STDOUT:
+		exporter, err = tracerstdout.NewExporter(stdout.Options{PrettyPrint: true})
```
Comment (honkfestival) [inline pkg/telemetry/telemetry.go:43]: If we don't `PrettyPrint`, does each JSON blob appear on its own line? If so, could we preserve that and maybe add a third provider called `PRETTY` that does the pretty printing?

JSON can also be pretty-printed downstream with e.g. `jq` and is easier to process if you know there's one object per line.

File: pkg/telemetry/telemetry.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+const PROMETHEUS = "prometheus"
+
+// InitTracer initializes the global trace provider
+func InitTracer() func() {
+	// Some providers require cleanup
+	cleanupFunc := func() {}
+
+	tracerProvider := os.Getenv("TRACER_PROVIDER")
+	if tracerProvider == "" {
+		log(nil, nil).Info("TRACER_PROVIDER not set, tracing will not be generated.")
```
Comment (honkfestival) [inline pkg/telemetry/telemetry.go:35]: 👍

File: pkg/telemetry/telemetry.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+		return cleanupFunc
+	}
+
+	var err error
+	switch metricProvider {
+	case STDOUT:
+		var pusher *push.Controller
+		pusher, err = metricstdout.InstallNewPipeline(metricstdout.Config{
+			Quantiles:   []float64{},
+			PrettyPrint: true,
```
Comment (honkfestival) [inline pkg/telemetry/telemetry.go:80]: (same question here)

File: pkg/telemetry/telemetry.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+	if err != nil {
+		log(nil, err).WithField("provider", metricProvider).Fatal("failed to initialize metric stdout exporter")
+	}
+
+	initSystemStatsObserver()
+
+	return cleanupFunc
+}
+
+// OpenTelemetryMiddleware adds monitoring around HTTP requests
```
Comment (honkfestival) [inline pkg/telemetry/telemetry.go:109]: Does this capture IP addresses?

Comment (honkfestival) [reply pkg/telemetry/telemetry.go:109]: (it looks like the answer is _no_ if your example above is complete)

File: CONTRIBUTING.md (hunk 4 above)
Comment (natalysheinin) [reply CONTRIBUTING.md:12]: ty

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -93,3 +116,45 @@ Finally, run:
...
+### Metrics
+
+Currently, the following options are supported for enabling Metrics:
+* standard output
+* prometheus
+
+Metrics can be enabled by setting the `METRIC_PROVIDER` variable to either `stdout`, or `prometheus`.
+
+If you want to use Prometheus, please see the additional configuration requirements below.
+
```
Comment (burke) [inline CONTRIBUTING.md:133]:
```suggestion
Note that logs are emitted to `stderr`, so with `stdout` mode, logs will be on `stderr` and metrics will be on `stdout`.
```

File: pkg/telemetry/telemetry.go (hunk 7 above)
Comment (genevieveluyt) [reply pkg/telemetry/telemetry.go:43]: Yes without the PrettyPrint the JSON is on one line. Separate provider for PRETTY sounds good to me 👍

File: pkg/telemetry/telemetry.go (hunk 10 above)
Comment (genevieveluyt) [reply pkg/telemetry/telemetry.go:109]: The example in the PR descriptions is indeed complete

File: cmd/monolith/main.go [hunk 12]
```diff
@@ -14,6 +15,12 @@ func main() {
 
 	log(nil, nil).Info("starting")
 
+	cleanupTracer := telemetry.InitTracer()
```
Comment (burke) [inline cmd/monolith/main.go:18]: Given that this is repeated in three places could we invoke it like:

```go
defer telemetry.Initialize().Cleanup()
```

File: pkg/telemetry/telemetry.go [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+var log = logger.New("telemetry")
+
+// Providers
+
+const STDOUT = "stdout"
+const PROMETHEUS = "prometheus"
+
```
Comment (burke) [inline pkg/telemetry/telemetry.go:27]: something like

```go
type traceMetricCleaner struct {
  tracer func()
  meter func()
}
type Cleanuper interface {
  Cleanup()
}
func (c *traceMetricCleaner) Cleanup() {
  c.tracer()
  c.meter()
}
func Initialize() Cleanuper {
  return &cleaner{tracer: InitTracer(), meter: InitMeter()}
}
```

File: pkg/telemetry/telemetry.go (hunk 10 above)
Comment (burke) [reply pkg/telemetry/telemetry.go:109]: It'd be worth adding a comment here something like:

"Be careful not to add anything here that captures personally-identify information such as IP addresses"

**Pull request merged**